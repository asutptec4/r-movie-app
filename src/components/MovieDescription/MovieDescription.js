import PropTypes from 'prop-types';
import React from 'react';

import { movie } from '../../types';
import { ratingToStr } from '../../utils/util-func';
import MoviePoster from '../MoviePoster/MoviePoster';
import './MovieDescription.scss';

const MovieDescription = ({ movie, closeButtonHandler }) => {
  return (
    <section className="movie-description-container">
      <div className="movie-description-buttons">
        <span className="logo app-logo">NetflixRoulette</span>
        <div className="icon-container" onClick={closeButtonHandler}>
          <span className="icon-glass"></span>
        </div>
      </div>
      {movie?.id && (
        <div className="movie-description">
          <MoviePoster imageUrl={movie.poster} className="movie-image" />
          <div className="description">
            <p className="movie-title">
              <span className="title">{movie.title}</span>
              <span className={'rating ' + ratingToStr(movie.rating)}>{movie.rating}</span>
            </p>
            <span className="movie-runtime">{movie.tagline}</span>
            <p className="movie-duration">
              <span>{movie.releaseYear}</span>
              <span>{movie.runtime} min</span>
            </p>
            <p className="movie-overview">{movie.overview}</p>
          </div>
        </div>
      )}
    </section>
  );
};

MovieDescription.defaultProps = {
  closeButtonHandler: () => {},
};

MovieDescription.propTypes = { closeButtonHandler: PropTypes.func, movie: movie };

export default MovieDescription;
