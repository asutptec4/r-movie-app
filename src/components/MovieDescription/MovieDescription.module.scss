@import '../../scss/index.scss';

$img-w: 280px;

.movieDescriptionContainer {
  padding: 1em;
  background-color: rgba(0, 0, 0, 0.5);

  .movieDescriptionButtons {
    margin-bottom: 1em;
    display: flex;
    flex-flow: row nowrap;
    justify-content: flex-end;

    .iconContainer {
      padding: 0.5em;
      cursor: pointer;

      .iconGlass {
        position: relative;
        display: inline-block;
        background-color: transparent;
        border-radius: 30px;
        height: 15px;
        width: 15px;
        border: 3px solid $item-color;

        &:after {
          content: '';
          height: 3px;
          width: 12px;
          background: $item-color;
          position: absolute;
          top: 15px;
          left: -12px;
          transform: rotate(145deg);
        }
      }
    }
  }

  .movieDescription {
    display: flex;
    flex-flow: row nowrap;
    padding: 0 2em;

    .movieImage {
      min-width: $img-w;
      width: $img-w;
      min-height: calc(4 / 3 * #{$img-w});
      height: calc(4 / 3 * #{$img-w});
    }

    .description {
      margin-left: 3em;
      max-height: calc(4 / 3 * #{$img-w});
      overflow: hidden;

      .movieTitle {
        margin-bottom: 0.5em;

        .title {
          font-size: $fs-large;
          margin-right: 1em;
        }

        .rating {
          font-size: $fs-logo;
          padding: 0.5em;
          border: 1px solid $label-color;
          border-radius: 50%;
        }

        .rating.good {
          color: yellowgreen;
        }

        .rating.norm {
          color: yellow;
        }

        .rating.bad {
          color: red;
        }
      }

      .movieRuntime {
        font-size: $fs-small;
      }

      .movieDuration {
        color: $item-color;
        font-size: $fs-logo;

        span {
          margin-right: 1em;
        }
      }
    }

    @media screen and (max-width: 800px) {
      .movieImage {
        display: none;
      }

      .description {
        margin-left: 0;
      }
    }
  }
}
