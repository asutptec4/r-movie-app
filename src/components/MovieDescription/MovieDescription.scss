@import '../../scss/index.scss';

$img-w: 280px;

.movie-description-container {
  .movie-description-buttons {
    margin-bottom: 1em;
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-between;

    .icon-container {
      padding: 0.5em;
      cursor: pointer;

      .icon-glass {
        position: relative;
        display: inline-block;
        background-color: transparent;
        border-radius: 30px;
        height: 15px;
        width: 15px;
        border: 3px solid $item-color;

        &:after {
          content: '';
          height: 3px;
          width: 12px;
          background: $item-color;
          position: absolute;
          top: 15px;
          left: -12px;
          transform: rotate(145deg);
        }
      }
    }
  }

  .movie-description {
    display: flex;
    flex-flow: row nowrap;
    padding: 0 2em;

    .movie-image {
      padding: 10px;
      width: $img-w;
      height: calc(4 / 3 * #{$img-w});
    }

    .description {
      margin-left: 3em;
      max-height: calc(4 / 3 * #{$img-w});
      overflow: hidden;

      .movie-title {
        margin-bottom: 0.5em;

        .title {
          font-size: $fs-large;
          margin-right: 1em;
        }

        .rating {
          font-size: $fs-logo;
          padding: 0.5em;
          border: 1px solid $label-color;
          border-radius: 50%;
        }

        .rating.good {
          color: yellowgreen;
        }

        .rating.norm {
          color: yellow;
        }

        .rating.bad {
          color: red;
        }
      }

      .movie-runtime {
        font-size: $fs-small;
      }

      .movie-duration {
        color: $item-color;
        font-size: $fs-logo;

        span {
          margin-right: 1em;
        }
      }
    }

    @media screen and (max-width: 800px) {
      .movie-image {
        display: none;
      }

      .description {
        margin-left: 0;
      }
    }
  }
}
