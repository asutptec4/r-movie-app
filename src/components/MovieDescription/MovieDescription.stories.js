import React from 'react';

import MovieDescription from './MovieDescription';

export default {
  title: 'MovieApp/MovieDescription',
  component: MovieDescription,
  parameters: { actions: { argTypesRegex: '^close.*' } },
};

const Template = (args) => <MovieDescription {...args} />;

export const Base = Template.bind({});
const movie = {
  genres: ['Science Fiction', 'Action', 'Adventure'],
  id: 424785,
  overview: 'Plot unknown.',
  poster: 'https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg',
  rating: '8.0',
  releaseDate: '2019-06-26',
  releaseYear: '2019',
  runtime: 120,
  subtitle: '',
  title: 'Transformers 7',
};
Base.args = {
  movie,
};
