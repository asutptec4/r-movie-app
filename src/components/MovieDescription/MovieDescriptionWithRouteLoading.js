import React, { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';

import { MOVIES_PATH, NOT_FOUND_PATH } from '../../constant';
import WithLoading from '../../hoc/WithLoading';
import {
  fetchMovie,
  resetIsNotFound,
  resetMovie,
  selectDetailMovie,
  selectLoading,
  selectNotFound,
} from '../../reducers/detailsSlice';
import MovieDescription from './MovieDescription';

const MovieDescriptionWithLoading = WithLoading(MovieDescription);

const MovieDescriptionWithRouteLoading = () => {
  const { movieId } = useParams();
  const history = useHistory();

  const detailMovie = useSelector(selectDetailMovie);
  const isLoading = useSelector(selectLoading);
  const isNotFound = useSelector(selectNotFound);
  const dispatch = useDispatch();

  const handleClose = useCallback(() => {
    history.push(MOVIES_PATH);
    dispatch(resetMovie());
  }, []);

  useEffect(() => {
    window.scrollTo(0, 0);
    dispatch(fetchMovie(movieId));
  }, [movieId]);

  useEffect(() => {
    if (isNotFound) {
      dispatch(resetIsNotFound());
      history.replace(NOT_FOUND_PATH);
    }
  }, [isNotFound]);

  return (
    <MovieDescriptionWithLoading
      movie={detailMovie}
      closeButtonHandler={handleClose}
      isLoading={isLoading}
    ></MovieDescriptionWithLoading>
  );
};

export default MovieDescriptionWithRouteLoading;
