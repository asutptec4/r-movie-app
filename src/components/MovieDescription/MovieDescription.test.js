import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import MovieDescription from './MovieDescription';

jest.mock('../MoviePoster/MoviePoster', () => 'MoviePosterComponent');

describe('<MovieDescription />', () => {
  const closeButtonHandler = jest.fn();
  const defaultProps = {
    movie: {
      genres: ['Science Fiction', 'Action', 'Adventure'],
      id: 424785,
      overview: 'Plot unknown.',
      poster: 'https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg',
      rating: '8.0',
      releaseDate: '2019-06-26',
      releaseYear: '2019',
      runtime: 120,
      subtitle: '',
      title: 'Transformers 7',
    },
    closeButtonHandler,
  };

  test('should render movie description', () => {
    const { container } = render(<MovieDescription {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('should invoke closeButtonHandler on glass icon click', () => {
    const { container } = render(<MovieDescription {...defaultProps} />);
    fireEvent.click(container.firstChild.firstChild.firstChild);
    expect(closeButtonHandler).toHaveBeenCalled();
  });
});
