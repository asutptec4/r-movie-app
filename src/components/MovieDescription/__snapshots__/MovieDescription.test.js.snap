// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDescription /> should render movie description 1`] = `
<div>
  <section
    class="movie-description-container"
  >
    <div
      class="movie-description-buttons"
    >
      <div
        class="icon-container"
      >
        <span
          class="icon-glass"
        />
      </div>
    </div>
    <div
      class="movie-description"
    >
      <div
        class="movie-image"
      >
        <moviepostercomponent
          imageurl="https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg"
        />
      </div>
      <div
        class="description"
      >
        <p
          class="movie-title"
        >
          <span
            class="title"
          >
            Transformers 7
          </span>
          <span
            class="rating good"
          >
            8.0
          </span>
        </p>
        <span
          class="movie-runtime"
        />
        <p
          class="movie-duration"
        >
          <span>
            2019
          </span>
          <span>
            120
             min
          </span>
        </p>
        <p
          class="movie-overview"
        >
          Plot unknown.
        </p>
      </div>
    </div>
  </section>
</div>
`;
